source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+  } else {
+    value = constant->AsIntConstant()->GetU32BitValue();
+  }
+
+  return CreateConstant(value);
+}
+
+// Handles both addition and subtraction. If the |sub| flag is set then the
+// addition will be op1+(-op2) otherwise op1+op2.
+SENode* ScalarEvolutionAnalysis::AnalyzeAddOp(const ir::Instruction* add,
+                                              bool sub) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+  std::unique_ptr<SENode> add_node{new SEAddNode(this)};
+
+  add_node->AddChild(
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(0))));
+
+  SENode* op2 =
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(1)));
+
+  // To handle subtraction we wrap the second operand in a unary negation node.
+  if (sub) {
+    op2 = CreateNegation(op2);
+  }
+  add_node->AddChild(op2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzePhiInstruction(
+    const ir::Instruction* phi) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  // Get the basic block this instruction belongs to.
+  ir::BasicBlock* basic_block =
+      context_->get_instr_block(const_cast<ir::Instruction*>(phi));
+
+  // And then the function that the basic blocks belongs to.
+  ir::Function* function = basic_block->GetParent();
+
+  // Use the function to get the loop descriptor.
+  ir::LoopDescriptor* loop_descriptor = context_->GetLoopDescriptor(function);
+
+  // We only handle phis in loops at the moment.
+  if (!loop_descriptor) return CreateCantComputeNode();
+
+  // Get the innermost loop which this block belongs to.
+  ir::Loop* loop = (*loop_descriptor)[basic_block->id()];
+  if (!loop) return CreateCantComputeNode();
+
+  std::unique_ptr<SERecurrentNode> phi_node{new SERecurrentNode(this, loop)};
+  instruction_map_[phi] = phi_node.get();
+
+  for (uint32_t i = 0; i < phi->NumInOperands(); i += 2) {
+    uint32_t value_id = phi->GetSingleWordInOperand(i);
+    uint32_t incoming_label_id = phi->GetSingleWordInOperand(i + 1);
+
+    ir::Instruction* value_inst = def_use->GetDef(value_id);
+    SENode* value_node = AnalyzeInstruction(value_inst);
+
+    // Both operands must be loop invariant.
+    if (!IsLoopInvariant(loop, value_node)) {
+      return CreateCantComputeNode();
+    }
+
+    if (incoming_label_id == loop->GetPreHeaderBlock()->id()) {
+      phi_node->AddOffset(value_node);
+    } else if (incoming_label_id == loop->GetLatchBlock()->id()) {
+      // Assumed to be in the form of step + phi.
+      SENode* constant_node = nullptr;
+      SENode* phi_operand = nullptr;
+      SENode* operand_1 = value_node->GetChild(0);
+      SENode* operand_2 = value_node->GetChild(1);
+
+      // Find which node is the step term.
+      if (!operand_1->AsSERecurrentNode())
+        constant_node = operand_1;
+      else if (!operand_2->AsSERecurrentNode())
+        constant_node = operand_2;
```

Comment:
From the name of the variable, it seems like you assume this is a constant, but it is not a constant for something like:

```
for( int i = 0; i < n; i+= step) {/*step does not change*/ }
```

Also you do not see to check that `constant_node` is even loop invariant.

How do you handle: 
```
int step = 0;
for( int i = 0; i < n; i += step) {step++;}
```

Reply:
The first case will work. Updated the names to reflect that. The second case we, now, don't support